qiskit/transpiler/target.py

```diff
@@ -0,0 +1,558 @@
+        cx_props = {
+            (0,1): InstructionProperties(length=5.23e-7, error=0.00098115),
+            (1,0): InstructionProperties(length=4.52e-7, error=0.00132115),
+        }
+        gmap.add_instruction(CXGate(), cx_props)
+
+    The intent of the ``Target`` object is to inform Qiskit's compiler about
+    the constraints of a particular backend so the compiler can compile an
+    input circuit to something that works and is optimized for a device.
+
+    Each instruction in the Target is indexed by a unique string name that uniquely
+    identifies that instance of an :class:`~qiskit.circuit.Instruction` object in
+    the Target. There is a 1:1 mapping between a name and an
+    :class:`~qiskit.circuit.Instruction` instance in the target and each name must
+    be unique. By default the name is the :attr:`~qiskit.circuit.Instruction.name`
+    attribute of the instruction, but can be set to anything. This lets a single
+    target have multiple instances of the same instruction class with different
+    parameters. For example, if a backend target has two instances of an
+    :class:`~qiskit.circuit.library.RXGate` one is parameterized over any theta
+    while the other is tuned up for a theta of pi/6 you can add these by doing something
+    like::
+
+        import math
+
+        from qiskit.transpiler import Target, InstructionProperties
+        from qiskit.circuit.library import RXGate
+        from qiskit.circuit import Parameter
+
+        target = Target()
+        theta = Parameter('theta')
+        rx_props = {
+            (0,): InstructionProperties(length=5.23e-8, error=0.00038115),
+        }
+        target.add_instruction(RXGate(theta), rx_props)
+        rx_30_props = {
+            (0,): InstructionProperties(length=5.23e-6, error=1.2e-5
+        }
+        target.add_instruction(RXGate(math.pi / 6), rx_30_props, name='rx_30')
+
+    Then in the ``target`` object accessing by ``rx_30`` will get the fixed
+    angle :class:`~qiskit.circuit.library.RXGate` while ``rx`` will get the
+    parameterized :class:`~qiskit.circuit.library.RXGate`.
+
+    .. note::
+
+        This class assumes that qubit indices start at 0 and are a contiguous
+        set if you want a submapping the bits will need to be reindexed in
+        a new object
+
+    .. note::
+
+        This class is designed to be additive only for gates, qargs, and qubits.
+        If you need to remove one of these the best option is to iterate over
+        an existing object and create a new subset (or use one of the methods
+        to do this). The object internally caches different views and these
+        would potentially be invalidated by removals.
+    """
+
+    __slots__ = (
+        "num_qubits",
+        "_gate_map",
+        "_gate_name_map",
+        "_qarg_gate_map",
+        "description",
+        "_coupling_graph",
+        "_instruction_durations",
+        "_instruction_schedule_map",
+        "dt",
+    )
```

Comment:
Transpiler doesn't use timestamp, however, when we did some experiments, we would need qubit T1/T2/frequency/anharmonicity/gate error information later to write a paper. Usually we access these information with `backend.properties(date_time)`. Or, some other researchers are interested in some statistical analysis of the trend of qubit properties (in this case they are interested in timestamp itself). Note that only backend properties information provide these logs, so something like `target.reflesh(date_time)` doesn't work (because we cannot retrieve past schedules).

Reply:
I see, that's a valid point and something I was helping people debug yesterday on an offline device they want to test with the properties from a device back in July and were having issues generating things consistently for transpilation. The issue I'm thinking of with this though is that from the terra perspective it has no way to know how to fetch information from a past date because that's part of the provider specific api (for example in the IBM case it's knowing how to request the backend properties via the iqx api for a specific date) but other providers have different mechanisms if one at all. Right now the ibm provider method is a custom thing for it only (as it's the only qiskit provider that exposes historical data) their overloaded `properties()` method optionally takes a date time. I'm thinking the ibm provider could have a similar thing when it migrates to backend v2, something like `backend.use_target_from(date_time)` which updates the target returned by `backend.target` to use properties from the provided date. Do you think this is a good solution for this? I can also add it as an optional method to the `BackendV2` class so it's an option for other providers in a standardized way (if any begin to offer a similar feature)